### ISSUE TITLE
how to deal with padding token

### ISSUE THREAD

**Issue by Earthones** [ArcInstitute/evo2#31]
Hi, when there are some different length sequences, which token should I adopt to padding. I have read the code of tokenizer.py, but there is no token to play this role, and if no padding word, how to deal with different sequence training for you? 
thank you!

**Comment by Sharique9022** [comment]
For handling different-length sequences, you should use a padding token. If your tokenizer does not have a predefined padding token, you can set one manually, e.g., tokenizer.pad_token = tokenizer.eos_token.

Additionally, consider:

Using padding=True during tokenization to ensure uniform sequence length.
Applying attention masks to differentiate real tokens from padding.
Ignoring padding tokens in loss calculation using ignore_index=tokenizer.pad_token_id in CrossEntropyLoss.
Using dynamic padding with PyTorch’s DataCollatorWithPadding for efficient batching.
Let me know if you need further clarification

**Comment by Earthones** [comment]
another problem is why embedding representation use the 28 layer's mlp output at readme, rather the 31 layer's (last layer),
    import torch
    from evo2 import Evo2
    
    evo2_model = Evo2('evo2_7b')
    
    sequence = 'ACGT'
    input_ids = torch.tensor(
        evo2_model.tokenizer.tokenize(sequence),
        dtype=torch.int,
    ).unsqueeze(0).to('cuda:0')
    
    layer_name = 'blocks.28.mlp.l3'
    
    outputs, embeddings = evo2_model(input_ids, return_embeddings=True, layer_names=[layer_name])
    
    print('Embeddings shape: ', embeddings[layer_name].shape)

**Comment by Earthones** [comment]
ok, I know. So when you are training evo2 model, padding = 1 is you padding token? following code comes from tokenizer.py 
class CharLevelTokenizer(AbstractTokenizer):
    """Character Level Tokenizer"""
    def __init__(self, vocab_size):
        name = "CharLevelTokenizer"
        super().__init__(name)
        self._vocab_size = vocab_size
        self.eod_id = 0
        self.eos_id = 0
        self.pad_id = 1
        ...
